rfc/bh-rfc-2.md

```diff
@@ -0,0 +1,114 @@
+---
+bh-rfc: 2
+title: Early Return Pattern
+authors: |
+    [Holms, Alyx](aholms@specterops.io)
+    [Pomeroy, Kaleb](kpomeroy@specterops.io)
+status: DRAFT
+created: 2025-04-15
+---
+
+# Early Return Pattern
+
+Go encourages keeping blocks small, variables scoped tightly, and failure paths easy to follow. While Inline Initializer Statements help with scoping, they should not be combined with chained nesting patterns, which obscure control flow and increase complexity. Nesting hides the success case in the deepest block, when that’s actually the flow we care about most.
+
+## 1. Overview
+
+This proposal recommends adopting the Early Return Pattern as the default style for handling sequential workflows and error propagation in Go. Early returns flatten control flow, make error handling explicit, and align with Go’s idiomatic practices as described in Effective Go. This change improves readability, reduces indentation depth, and creates more maintainable and composable functions.
```

Comment:
This is an alright overview currently, but some of these things are true with the existing pattern as well. Error handling is still explicit in the nesting version, you're still early returning (implicitly, though it can also be made explicit, it just looks weird), and really control flow isn't that much more  flat, because again it's semantically not flat, but visually it's pretty flat.

Maintenance is something that could be improved by moving away from the chained if/else pattern, but we need more concrete information on _why_ that is and what specific footguns we run into today that would be easier to avoid with this approach.